mailpoet/lib/Cron/Workers/SendingQueue/SendingQueue.php

```diff
@@ -160,13 +166,34 @@ private function processSending(SendingTask $queue, int $timer): void {
       return;
     }
 
+    $campaignId = null;
+
+    $afterPreProcessFilter = function(array $renderedNewsletters, NewsletterEntity $renderedNewsletterEntity) use ($newsletterEntity, &$campaignId) {
+      if ($newsletterEntity !== $renderedNewsletterEntity || !isset($renderedNewsletters['text'])) {
+        return;
+      }
+      $textVersion = $renderedNewsletters['text'];
+      $campaignId = $this->calculateCampaignId($newsletterEntity, $textVersion);
+      return $renderedNewsletters;
+    };
+
+    // This filter fires during preProcessNewsletter, after some initial rendering but before any shortcodes are replaced.
+    $this->wp->addFilter('mailpoet_sending_newsletter_render_after_pre_process', $afterPreProcessFilter, 10, 2);
+
     // pre-process newsletter (render, replace shortcodes/links, etc.)
     $newsletterEntity = $this->newsletterTask->preProcessNewsletter($newsletterEntity, $queue);
+
+    $this->wp->removeFilter('mailpoet_sending_newsletter_render_after_pre_process', $afterPreProcessFilter);
```

Comment:
It feels like there is probably a better way to handle this, but I haven't been able to come up with one. 

I originally thought I could have the calculateCampaignId method do some rendering itself, but if you have the renderer render the same newsletter twice it may not produce the same results, e.g. it will skip ALC posts that were previously rendered.

Ultimately the idea of using this filter appealed to me because I think the newsletter is in just the right state. Shortcodes haven't been replaced yet, which I think is good for what we're trying to achieve.

I could have done some refactoring, but lots of this code is going to need refactoring when we convert it to doctrine, and I'm honestly nervous about changing too much in any of this code.

That being said, I'm very much open to better ideas!

Reply:
This looks good to me, the only "why don't you just" question I'll try will be this one:

Why not implementing this within the `preProcessNewsletter` function that you're wrapping here via shortcodes? In other words — can we put the code to the place from which this hook is invoked? Or would that be a worse spot to place this?

Reply:
Such a simple question and yet something that I didn't properly consider!

Mainly I was thinking that it only makes sense to compute the value when we're sure a send is about to occur, and I (incorrectly, I believe) assumed that `preProcessNewsletter` might be used in other situations.

I think [this](https://github.com/mailpoet/mailpoet/blob/MAILPOET-5034%2Fgenerate_campaign_ids_for_new_email_campaigns/mailpoet/lib/Newsletter/NewsletterSaveController.php#L415-L415) is the only other usage, and long story short I think by not including this campaignId code directly in `preProcessNewsletter` we could end up sending the wrong campaignId for newsletters that get scheduled and then the content gets updated.

Thanks so much for pointing this out! I'll move the code over there and triple check that it won't cause any issues.